src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -1638,14 +2054,93 @@ def __init__(self, *args, **kwargs):
             'CA_CREATE_INTERMEDIATE': self.__create_intermediate_ca,
         }
 
+    @accepts()
+    async def profiles(self):
+        return self.PROFILES
+
+    @periodic(86400, run_on_start=True)
+    @private
+    async def crl_generation(self):
+        await self.middleware.call('service.start', 'ssl')
+
     # HELPER METHODS
 
+    @private
+    async def revoke_ca_chain(self, ca_id):
+        chain = await self.get_ca_chain(ca_id)
+        for cert in chain:
+            datastore = f'system.certificate{"authority" if cert["cert_type"] == "CA" else ""}'
+            await self.middleware.call(
+                'datastore.update',
+                datastore,
+                cert['id'], {
+                    'revoked': True,
+                    'revoked_date': datetime.datetime.utcnow()
+                },
+                {'prefix': self._config.datastore_prefix}
+            )
+
+    @private
+    async def get_ca_chain(self, ca_id):
+        certs = list(
+            map(
+                lambda item: dict(item, cert_type='CERTIFICATE'),
+                await self.middleware.call(
+                    'datastore.query',
+                    'system.certificate',
+                    [['signedby', '=', ca_id]],
+                    {'prefix': self._config.datastore_prefix}
+                )
+            )
+        )
+
+        for ca in await self.middleware.call(
+            'datastore.query',
+            'system.certificateauthority',
+            [['signedby', '=', ca_id]],
+            {'prefix': self._config.datastore_prefix}
+        ):
+            certs.extend((await self.get_ca_chain(ca['id'])))
+
+        ca = await self.middleware.call(
+            'datastore.query',
+            'system.certificateauthority',
+            [['id', '=', ca_id]],
+            {'prefix': self._config.datastore_prefix, 'get': True}
+        )
+        ca.update({'cert_type': 'CA'})
+
+        certs.append(ca)
+        return certs
+
     @private
     async def validate_common_attributes(self, data, schema_name):
         verrors = ValidationErrors()
 
         await _validate_common_attributes(self.middleware, data, verrors, schema_name)
 
+        if not data['cert_extensions']['BasicConstraints']['enabled']:
```

Comment:
How can you be sure these cert_extensions have these keys?

cert_extensions does not seem to be marked as required, neither each of its keys.

Reply:
But we do set their defaults so even if they aren't marked as required, we have the default dictionary with specified defaults

Reply:
How about API calls, where user sents a payload with only cert_extensions without the needed keys?